### PR TITLE
Updated test harness

### PR DESCRIPTION
**Pull request by prafulmehrotra**
Updated script to keep track of ssh keys created per region as part of test harness and to delete them when the tests are done.

### CODE REVIEW

File: bin/test.py [hunk 1]
```diff
@@ -204,14 +213,16 @@ def wait_for_stacks(stacks, create):
...
     """
+    key_pairs=[]
     for region in get_regions(session.client('ec2', region_name='us-east-1', config=config)):
 
         print 'Ensure key pair in region: {}'.format(region)
 
         ec2_client = session.client('ec2', region_name=region, config=config)
         key_name = '{}{}'.format(KEY_PAIR_PREFIX, region)
         if not has_key_pair(ec2_client, key_name):
-            store_key_locally(key_name, create_key_pair(ec2_client, key_name))
```
Comment (rnzsgh) [inline bin/test.py]: This function is not in use and should be removed.

File: bin/test.py [hunk 2]
```diff
@@ -224,6 +235,17 @@ def ensure_foundation(session, config):
         # Copy the latest version of the file to the bucket
         update_sample_app(s3_client, app_bucket_name, TEST_APP_SOURCE_BUCKET, TEST_PYTHON_APP_KEY)
 
+    return key_pairs    
+
+def remove_keypairs(session,config,key_pairs):
```
Comment (rnzsgh) [inline bin/test.py]: Add spaces between parameters after ,

E.g.,

(foo, bar, test)

File: bin/test.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 
 TEST_PYTHON_APP_KEY='eb-python-flask.zip'
 
+TEMPLATE_BUCKET='awslabs-startup-kit-templates-deploy-v2'
```
Comment (rnzsgh) [inline bin/test.py]: Reuse in the TEMPLATE_URL_PREFIX?

File: bin/test.py [hunk 4]
```diff
@@ -204,14 +213,16 @@ def wait_for_stacks(stacks, create):
...
     for region in get_regions(session.client('ec2', region_name='us-east-1', config=config)):
 
         print 'Ensure key pair in region: {}'.format(region)
 
         ec2_client = session.client('ec2', region_name=region, config=config)
         key_name = '{}{}'.format(KEY_PAIR_PREFIX, region)
         if not has_key_pair(ec2_client, key_name):
-            store_key_locally(key_name, create_key_pair(ec2_client, key_name))
+            create_key_pair(ec2_client, key_name)
+            key_pairs.append({"Region":region,"KeyName":key_name})
```
Comment (rnzsgh) [inline bin/test.py]: You need to store the key value to if the key is to be used.

File: bin/test.py [hunk 5]
```diff
@@ -224,6 +235,17 @@ def ensure_foundation(session, config):
...
+    return key_pairs    
+
+def remove_keypairs(session,config,key_pairs):
+    """ Remove key-pairs created as part of test harness
+    """
+    for key_pair in key_pairs:
+
+        ec2_client = session.client('ec2', region_name=key_pair['Region'], config=config)
+        delete_key_pair(ec2_client,key_pair['KeyName'])
+        print 'Deleted keypair: {} in region: {}'.format(key_pair['KeyName'],key_pair['Region'])
```
Comment (rnzsgh) [inline bin/test.py]: Add spaces between the params

File: bin/test.py [hunk 6]
```diff
@@ -204,14 +213,16 @@ def wait_for_stacks(stacks, create):
 def ensure_foundation(session, config):
     """ Make sure we have everything we need in place to run the stacks
     """
+    key_pairs=[]
     for region in get_regions(session.client('ec2', region_name='us-east-1', config=config)):
 
         print 'Ensure key pair in region: {}'.format(region)
 
         ec2_client = session.client('ec2', region_name=region, config=config)
         key_name = '{}{}'.format(KEY_PAIR_PREFIX, region)
```
Comment (rnzsgh) [inline bin/test.py]: Can you add the epoch time in seconds after the key pair prefix?

E.g.,
key-name-prefix-345435345344-us-east-1

File: bin/test.py (hunk 5 above)
Comment (rnzsgh) [inline bin/test.py]: Add spaces in the params

File: bin/test.py (hunk 1 above)
Comment (prafulmehrotra) [reply bin/test.py]: Removed function.

**Comment by prafulmehrotra** [comment]
Ryan - fixed all issues except epoch. Will add that in the next commit as fixing that breaks the code in a few places. As for explicitly defining TEMPLATE_BUCKET, I do this so as to get around the default condition. When you are testing the wrapper templates, either in your own test bucket or the live bucket, the nested stacks that are created use templates in the live bucket by default unless you explicitly define your cloudformation bucket in the wrapper template.

**Pull request merged**